Source/AwsCommonRuntimeKit/http/HttpProxyOptions.swift

```diff
@@ -3,43 +3,36 @@
 import AwsCHttp
 import AwsCCommon
 
-// Todo: try to find a better for a persistent byte cursor
-public class HttpProxyOptions {
-    private let allocator: Allocator
-    private var rawValue: UnsafeMutablePointer<aws_http_proxy_options>
-    private var _basicAuthUsername: AWSStringByteCursor?
-    private var _basicAuthPassword: AWSStringByteCursor?
-    private var _hostName: AWSStringByteCursor?
-
+public class HttpProxyOptions: CStruct {
     public var authType: HttpProxyAuthenticationType = .none
     public var basicAuthUsername: String?
     public var basicAuthPassword: String?
     public var hostName: String
     public var port: UInt16
     public var tlsOptions: TlsConnectionOptions?
 
-    public init(hostName: String, port: UInt16, allocator: Allocator = defaultAllocator) {
-        self.allocator = allocator
-        self.rawValue = allocator.allocate(capacity: 1)
+    public init(hostName: String, port: UInt16) {
         self.hostName = hostName
         self.port = port
     }
 
-    func getRawValue() -> UnsafeMutablePointer<aws_http_proxy_options> {
-        _basicAuthUsername = AWSStringByteCursor(basicAuthUsername ?? "", allocator: allocator)
-        _basicAuthPassword = AWSStringByteCursor(basicAuthPassword ?? "", allocator: allocator)
-        _hostName = AWSStringByteCursor(hostName, allocator: allocator)
-        rawValue.pointee.port = port
-        rawValue.pointee.host = _hostName!.byteCursor
-        rawValue.pointee.auth_username = _basicAuthUsername!.byteCursor
-        rawValue.pointee.auth_password = _basicAuthPassword!.byteCursor
-        rawValue.pointee.tls_options = UnsafePointer(tlsOptions?.rawValue)
-        rawValue.pointee.auth_type = authType.rawValue
-
-        return rawValue
-    }
+    typealias RawType = aws_http_proxy_options
+    func withCStruct<Result>(shutdownOptions _: aws_shutdown_callback_options? = nil,
+                             _ body: (aws_http_proxy_options) -> Result
+    ) -> Result {
+        var cProxyOptions = aws_http_proxy_options()
+        cProxyOptions.port = port
+        cProxyOptions.auth_type = authType.rawValue
+        cProxyOptions.tls_options = UnsafePointer(tlsOptions?.rawValue)
 
-    deinit {
-        allocator.release(rawValue)
+        return withByteCursorFromStrings(basicAuthUsername ?? "",
+                                         basicAuthPassword ?? "",
+                                         hostName) {
+            userNamePointer, passwordPointer, hostPointer in
+            cProxyOptions.host = hostPointer
+            cProxyOptions.auth_username = userNamePointer
+            cProxyOptions.auth_password = passwordPointer
+            return body(cProxyOptions)
+        }
```

Comment:
The problem with just having this function without any protocol was that ProxyOptions are nullable. So 
`proxyOptions.withCStruct() {cStruct in }` wouldn't work if proxyOptions is null. Another solution I tried was to make it a static function at https://github.com/awslabs/aws-crt-swift/pull/88/commits/0459fd145dc2e95fbd2ff882723743736c168607. That approach can also work but there is no relation between different structs which is more flexible but less consistent.

Reply:
I am not sure if this complexity is worth the consistency. The main problem with static function approach I think is that we cannot create a generic helper function to avoid nested closures.